vpr/src/place/move_utils.cpp

```diff
@@ -28,6 +28,52 @@ void set_placer_breakpoint_reached(bool flag) {
     f_placer_breakpoint_reached = flag;
 }
 
+/**
+ * @brief Adjust the search range based on the block type and constraints
+ * 
+ * If the block is an IO block, we expand the search range to include all blocks in the column
+ * We found empirically that this is a good strategy for IO blocks given they are located in
+ * the periphery for most FPGA architectures
+ * 
+ * @param block_type The type of the block to move
+ * @param block_id The block ID of the moving block
+ * @param search_range The search range to adjust
+ * @param delta_cx The delta x of the search range
+ * @param to_layer_num The layer that the block is moving to
+ * 
+ * @return true if the search range was adjusted, false otherwise
+ */
+static bool adjust_search_range(t_logical_block_type_ptr block_type,
+                                ClusterBlockId block_id,
+                                t_bb& search_range,
+                                int& delta_cx,
+                                int to_layer_num) {
+
+    auto block_constrained = is_cluster_constrained(block_id);
+
+    if (block_constrained) {
+        bool intersect = intersect_range_limit_with_floorplan_constraints(block_id,
+                                                                          search_range,
+                                                                          delta_cx,
+                                                                          to_layer_num);
+        if (!intersect) {
+            return false;
+        }
+    }
+
+    if (is_io_type(block_type) && !block_constrained) {
+        /* We empirically found that for the IO blocks,
+         * Given their sparsity, we expand the y-axis search range 
+         * to include all blocks in the column
+         */
+        const t_compressed_block_grid& compressed_block_grid = g_vpr_ctx.placement().compressed_block_grids[block_type->index];
```

Comment:
You could make this code work well for area-IO (e.g. stratix 10 where IOs are in columns) and for other sparse blocks by just checking if the number of blocks in the compressed columns is less than some threshold.  I suspect you don't actually need to check if it is an IO block -- IOs are just an example of this kind of pattern.

(I still like the utility to check if a logical block type is an IO though).

Reply:
I actually tried that a while ago by not specifying the block type and simply expanding the search range if the number of blocks in the chosen column was less than 3. However, after making that change, in the example I mentioned earlier, we encountered the same issue where the IO blocks became scattered between the top and bottom edges.